returnn/perplexity.py

```diff
@@ -0,0 +1,93 @@
+__all__ = ["ReturnnCalculatePerplexityJob"]
+
+import shutil
+import subprocess as sp
+from typing import Union
+
+from sisyphus import Job, Task, setup_path, tk
+
+import i6_core.util as util
+
+from .config import ReturnnConfig
+from .training import PtCheckpoint, Checkpoint
+
+Path = setup_path(__package__)
+
+
+class ReturnnCalculatePerplexityJob(Job):
+    """
+    Calculates the perplexity of a language model trained in RETURNN
+    on an evaluation data set
+    """
+
+    def __init__(
+        self,
+        returnn_config: ReturnnConfig,
+        returnn_model: Union[PtCheckpoint, Checkpoint],
+        eval_dataset: tk.Path,
+        *,
+        log_verbosity: int = 3,
+        returnn_root: tk.Path,
+        returnn_python_exe: tk.Path,
+    ):
+        returnn_config.config.pop("train")
+        returnn_config.config.pop("dev")
+        returnn_config.config["eval_datasets"] = {"eval": eval_dataset}
+
+        # TODO verify paths
+        if isinstance(returnn_model, PtCheckpoint):
+            model_path = returnn_model.path
+            self.add_input(returnn_model.path)
```

Comment:
Are you sure??
Pt/Checkpoint are normal Python classes which I think are not covered by the `extract_paths` from sisyphus https://github.com/rwth-i6/sisyphus/blob/master/sisyphus/tools.py#L74
or am I missing something?

Reply:
> Are you sure??

never without actually testing ;)

but very confident because:
* in recognition we also just give the checkpoint object and it works
> are not covered by the extract_paths from sisyphus
* extract_paths should arrive in the last else and then call get_object_state which should then via get_members_ descend into the __dict__ of the Checkpoint object and return the underlying tk.Path object.

Reply:
ah yes. thanks for the explanation :)